src/fadt.rs

```diff
@@ -0,0 +1,185 @@
+use {AcpiError, PhysicalMapping};
+use sdt::SdtHeader;
+#[cfg(test)] use std::mem;
+
+#[repr(C, packed)]
+struct GenericAddress {
+    address_space: u8,
+    bit_width: u8,
+    bit_offset: u8,
+    access_size: u8,
+    address: u64,
+}
+
+impl GenericAddress {
+    #[cfg(test)]
+    pub(crate) fn make_testcase() -> GenericAddress {
+        GenericAddress {
+            address_space: 0 as u8,
+            bit_width: 0 as u8,
+            bit_offset: 0 as u8,
+            access_size: 0 as u8,
+            address: 0 as u64,
+        }
+    }
+}
+
+#[repr(C, packed)]
+pub struct Fadt {
+    // header
+    header: SdtHeader,
+    
+    firmware_ctrl: u32,
+    dsdt_address: u32,
+
+    // used in acpi 1.0; compatibility only, should be zero
+    reserved: u8,
+
+    preferred_pm_profile: u8,
+    sci_interrupt: u16,
+    smi_cmd_port: u32,
+    acpi_enable: u8,
+    acpi_disable: u8,
+    s4bios_req: u8,
+    pstate_control: u8,
+    pm1a_event_block: u32,
+    pm1b_event_block: u32,
+    pm1a_control_block: u32,
+    pm1b_control_block: u32,
+    pm2_control_block: u32,
+    pm_timer_block: u32,
+    gpe0_block: u32,
+    gpe1_block: u32,
+    pm1_event_length: u8,
+    pm1_control_length: u8,
+    pm2_control_length: u8,
+    pm_timer_length: u8,
+    gpe0_block_length: u8,
+    gpe1_block_length: u8,
+    gpe1_base: u8,
+    c_state_control: u8,
+    worst_c2_latency: u16,
+    worst_c3_latency: u16,
+    flush_size: u16,
+    flush_stride: u16,
+    duty_offset: u8,
+    duty_width: u8,
+    day_alarm: u8,
+    month_alarm: u8,
+    century: u8,
+    iapc_boot_arch: u16,
+    reserved2: u8, // must be 0
+    flags: u32,
+    reset_reg: GenericAddress,
+    reset_value: u8,
+    arm_boot_arch: u16,
+    fadt_minor_version: u8,
+    x_firmware_control: u64,
+    x_dsdt_address: u64,
+    x_pm1a_event_block: GenericAddress,
+    x_pm1b_event_block: GenericAddress,
+    x_pm1a_control_block: GenericAddress,
+    x_pm1b_control_block: GenericAddress,
+    x_pm2_control_block: GenericAddress,
+    x_pm_timer_block: GenericAddress,
+    x_gpe0_block: GenericAddress,
+    x_gpe1_block: GenericAddress,
+    sleep_control_reg: GenericAddress,
+    sleep_status_reg: GenericAddress,
+    hypervisor_vendor_id: u64,
+}
+
+impl Fadt {
+    pub fn validate(&self) -> Result<(), AcpiError> {
```

Comment:
`SdtHeader::validate` was updated in [this commit](https://github.com/rust-osdev/acpi/commit/f578070927d97bdf891c1c80bcdd51e2ebab4b3f), so this function could be replaced with `self.header.validate(b"FACP")` at the call-site. You'll probably need to rebase anyways, so might as well update this

Reply:
After a painfull rebase and merge (I think I fucked up somewhere haha), it is changed :smile:

Reply:
Also changed `==` to `!=` in the validate signature :p

Reply:
Oh wow; that's why we needs tests I guess :laughing:. And looks like you've included all of the changes again instead of rebasing, but it'll apparently still auto-merge, so I'm cool with that. Git is such as pain sometimes.

Reply:
Yeah I forgot to squash, will do it later so the log history won't be too ugly :p